src/Generator.cpp

```diff
@@ -97,29 +99,562 @@ Outputs compute_outputs(const Target &target,
+        if (next.first == 0 && next.second == 0) {
+            hi = mid;
+        } else {
+            lo = mid;
+            best = next;
+        }
+    }
+    
+    return {best.first * sign, best.second};
+}
+
+std::vector<Type> parse_halide_type_list(const std::string &types) {
+    const auto &e = get_halide_type_enum_map();
+    std::vector<Type> result;
+    for (auto t : split_string(types, ",")) {
+        auto it = e.find(t);
+        user_assert(it != e.end()) << "Type not found: " << t;
+        result.push_back(it->second);
+    }
+    return result;
+}
+
+template<typename T>
+T parse_scalar(const std::string &value) {
+    std::istringstream iss(value);
+    T t;
+    iss >> t;
+    user_assert(!iss.fail() && iss.get() == EOF) << "Unable to parse: " << value;
+    return t;
+}
+
 }  // namespace
 
-const std::map<std::string, Halide::Type> &get_halide_type_enum_map() {
-    static const std::map<std::string, Halide::Type> halide_type_enum_map{
-        {"bool", Halide::Bool()},
-        {"int8", Halide::Int(8)},
-        {"int16", Halide::Int(16)},
-        {"int32", Halide::Int(32)},
-        {"uint8", Halide::UInt(8)},
-        {"uint16", Halide::UInt(16)},
-        {"uint32", Halide::UInt(32)},
-        {"float32", Halide::Float(32)},
-        {"float64", Halide::Float(64)}
+class StubEmitter {
+public:
+    StubEmitter(std::ostream &dest, 
+                   const std::string &fully_qualified_name,
+                   const std::vector<Internal::GeneratorParamBase *>& generator_params,
+                   const std::vector<Internal::GeneratorInputBase *>& inputs,
+                   const std::vector<Internal::GeneratorOutputBase *>& outputs) 
+        : stream(dest), fully_qualified_name(fully_qualified_name), generator_params(filter_params(generator_params, false)), 
+          schedule_params(filter_params(generator_params, true)), inputs(inputs), outputs(outputs) {
+        internal_assert(!outputs.empty());
+    }
+
+    void emit();
+private:
+    std::ostream &stream;
+    const std::string fully_qualified_name;
+    const std::vector<Internal::GeneratorParamBase *> generator_params;
+    const std::vector<Internal::GeneratorParamBase *> schedule_params;
+    const std::vector<Internal::GeneratorInputBase *> inputs;
+    const std::vector<Internal::GeneratorOutputBase *> outputs;
+    int indent{0};
+
+    std::vector<Internal::GeneratorParamBase *> filter_params(const std::vector<Internal::GeneratorParamBase *> &in, 
+                                                              bool is_schedule_params) {
+        std::vector<Internal::GeneratorParamBase *> out;
+        for (auto p : in) {
+            if (p->name == "target") continue;
+            if (p->is_schedule_param() != is_schedule_params) continue;
+            out.push_back(p);
+        }
+        return out;
+    }
+
+    /** Emit spaces according to the current indentation level */
+    std::string ind();
```

Comment:
I don't think the savings of three characters here is worth the loss of clarity.